README.md
Puppet Facts Upload
===================

[![Build Status](https://travis-ci.org/Sharpie/puppet-facts_upload.svg?branch=master)](https://travis-ci.org/Sharpie/puppet-facts_upload)

This project re-implements the `puppet facts upload` command and part of the
associated `/puppet/v3/facts` endpoint that [was removed][PUP-2560] along with
the inventory service in Puppet 4.0. These pieces of functionality were useful
for keeping facts updated outside of catalog compilation.

[PUP-2560]: https://tickets.puppetlabs.com/browse/PUP-2560

Re-adding the command and endpoint enables two key use cases:

  - The ability to run agents off of cached catalogs. Facts become stale in this
    configuration as they are only updated when a fresh catalog is compiled.

  - The ability to submit updated facts at the end of an agent run. This allows
    facts to reflect any changes made during catalog application and keeps fact
    data as fresh as possible for PuppetDB queries.


## Requirements

This module has been tested against the following versions of Puppet Server and
Puppet Enterprise running on the following platforms:

  - **Versions:** Puppet Server 5.1.z and 5.2.z,
                  PE 2017.3.z and PE 2016.4.10 -- PE 2016.4.z
  - **Platforms:** RedHat 6, 7; Ubuntu 14.04, 16.04; SLES 11, 12

The `facts_upload` module is fairly sensitive to the Puppet Server and PE
version numbers as the code hooks into several internal APIs in both Puppet
Server and the `puppet_enterprise` module. Installation of the Java plugins
also requires manipulation of Puppet Server's Java classpath for versions
older than Puppet Server 5 or PE 2017.3. The module likely will not work if
your Puppet Server version or platform is not on the list above.

Puppet Agent support is much less restrictive. The module should work with Puppet
versions 4.y and 5.y.


## Usage

The `facts_upload` module currently provides one class: `facts_upload::server`.
This class should be applied to all nodes running the `puppetserver` or
`pe-puppetserver` services and will configure the Puppet Server to mount a
`/puppet/v3/facts` endpoint and configures `auth.conf` to allow agents to
upload facts for just their own certnames.

The module also provides some Ruby code which is pluginsynced to agents:

  - `lib/puppet/indirector/facts/rest.rb`
  - `lib/puppet/feature/facts_upload_builtin.rb`
  - `lib/puppet/face/facts/upload.rb`

Agents may run `puppet facts upload` once pluginsync has completed.


## Limitations

This module only provides a subset of the functionality that was removed in
[PUP-2560][PUP-2560]:

  - A `/puppet/v3/facts` API endpoint for Puppet Server that responds to PUT
    requests only.

  - A `Puppet::Node::Facts::Rest` indirector terminus that sends requests from
    agents to the `/puppet/v3/facts` API endpoint of the `server` the agent is
    configured to talk with.

  - A `puppet facts upload` face that fetches facts from Facter and then
    uploads them to the `/puppet/v3/facts` API endpoint of the `server` the
    agent is configured to talk with.

Notably, the following deprecated features have not been restored:

  - Functionality and REST API routes related to the Inventory Service.

  - Functionality and indirector terminii related to  the `store_configs`
    subsystem.


## Development

See the output of `rake -T` for development workflow tasks. More info will be
added to this section in a future release.


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [1.1.0] - 2018-11-14
### Added

  - The `facts_upload::server` class now supports PE 2016.4.10 -- 2016.4.latest
    in addition to PE 2017.3.

  - A pre-compiled JAR file is now checked into the git repository so that the
    module may be installed directly from git.

  - JAR plugin compatibility has been extended back to Puppet Server 2.6.


## [1.0.4] - 2018-11-03
### Fixed

  - The module can now be installed in environments using version 5.x
    of puppetlabs-stdlib or puppetlabs-concat. This requires updating
    the puppetlabs-puppet_authorization module to 0.5.0 or newer.

  - The module no longer triggers a "Could not autoload puppet/type/pe_ini_setting"
    error when Code Manager is enabled.

## [1.0.3] - 2018-04-19
### Fixed

  - The `facts_upload.jar` plugin will no longer activate when loaded by
    a Puppet Server version that the plugin is not compatible with.

  - The `facts_upload` module will no longer re-define the `puppet facts upload`
    command for Puppet versions 5.5 and higher.

## [1.0.2] - 2018-02-21
### Added

  - Print name of server receiving upload to messages logged by
    `puppet facts upload`.

### Fixed

  - Use SSL and server settings from the `agent` section of `puppet.conf
    when running `puppet facts upload`.

  - Use Puppet's `node_name_value` and `node_name_fact` settings to match
    behavior with `puppet agent` and `puppet apply`.


## [1.0.1] - 2017-12-06
### Fixed

  - Ruby patch code is only applied once per JRuby which prevents a
    "stack level too deep" error caused by applying the patch multiple
    times.


## [1.0.0] - 2017-12-05
### Added

  - A `/puppet/v3/facts` API endpoint for Puppet Server that responds to PUT
    requests only.

  - A `Puppet::Node::Facts::Rest` indirector terminus that sends requests from
    agents to the `/puppet/v3/facts` API endpoint of the `server` the agent is
    configure to talk with.

  - A `facts_upload::server` class that configures Puppet Server 5.1.z or
    PE 2017.3.z to mount the above.

  - A `puppet facts upload` face that fetches facts from Facter and then
    uploads them to the `/puppet/v3/facts` API endpoint of the `server` the
    agent is configured to talk with.

  - Tests. Tests. Tests.


[1.1.0]: https://github.com/Sharpie/puppet-facts_upload/compare/1.0.4...1.1.0
[1.0.4]: https://github.com/Sharpie/puppet-facts_upload/compare/1.0.3...1.0.4
[1.0.3]: https://github.com/Sharpie/puppet-facts_upload/compare/1.0.2...1.0.3
[1.0.2]: https://github.com/Sharpie/puppet-facts_upload/compare/1.0.1...1.0.2
[1.0.1]: https://github.com/Sharpie/puppet-facts_upload/compare/1.0.0...1.0.1
[1.0.0]: https://github.com/Sharpie/puppet-facts_upload/compare/5620a62...1.0.0


test/acceptance/tests/00_instal_module.rb
require 'json'

step 'Copy module to master VM' do
  metadata = JSON.parse(File.read('metadata.json'))
  module_tarball = "#{metadata['name']}-#{metadata['version']}.tar.gz"

  # We use this instead of `copy_module_to` from beaker-puppet as the
  # Rakefile contains logic for building a tarball that contains an
  # assembled JAR.
  master.do_scp_to("pkg/#{module_tarball}", "/tmp/#{module_tarball}", {})
  # The uninstall is here to ensure we load in a fresh copy of the
  # module if the tests are being re-run.
  on(master, puppet('module', 'uninstall', 'sharpie-facts_upload'),
    accept_all_exit_codes: true)
  on(master, puppet('module', 'install', "/tmp/#{module_tarball}"))

  apply_manifest_on(master, <<-EOM)
service{'puppetserver': ensure => running}
class{'facts_upload': }
EOM
end


test/acceptance/pre_suite/10_install_pe.rb
require 'yaml'
require 'beaker-pe'

test_name 'Install PE' do
  confine :to, :type => 'pe'

  step 'Install PE prereqs' do
    if master[:hypervisor] == 'docker'
      # TODO: Prereq for CentOS 7. Map out additonal PE master platforms:
      #  - CentOS 6
      #  - Ubuntu 16.04
      #  - Ubuntu 14.04
      #  - Sles 12
      #  - SLES 11
      hosts.each do |h|
        install_package(h, 'cronie')
      end
    end
  end

  step 'Install Puppet Enterprise' do
    answers = {'puppet_enterprise::puppet_master_host': '%{::trusted.certname}',
               'pe_install::puppet_master_dnsaltnames': ['%{::trusted.certname}',
                                                         "#{master.hostname}",
                                                         'puppet'],
               'puppet_enterprise::master::puppetserver::jruby_max_active_instances': 1,
               'puppet_enterprise::profile::master::java_args': {'Xmx': '384m',
                                                                 'Xms': '128m'},
               'puppet_enterprise::profile::puppetdb::java_args': {'Xmx': '128m',
                                                                   'Xms': '64m'},
               'puppet_enterprise::profile::console::java_args': {'Xmx': '96m',
                                                                  'Xms': '64m'},
               'puppet_enterprise::profile::orchestrator::java_args': {'Xmx': '64m',
                                                                       'Xms': '64m'}}

    # Ensure memory tuning persists on older PE versions.
    on(master, 'mkdir -p /etc/puppetlabs/code/environments/production/hieradata')
    create_remote_file(master,
                       '/etc/puppetlabs/code/environments/production/hieradata/common.yaml',
                       answers.to_yaml)
    on(master, 'mkdir -p /etc/puppetlabs/code/environments/production/data')
    create_remote_file(master,
                       '/etc/puppetlabs/code/environments/production/data/common.yaml',
                       answers.to_yaml)

    install_pe_on(hosts, answers: answers)
    create_remote_file(master, '/etc/puppetlabs/puppet/autosign.conf', "*\n")
    on(master, 'chown pe-puppet /etc/puppetlabs/puppet/autosign.conf')
  end
end


test/acceptance/tests/00_install_module.rb
require 'json'
require 'rubygems/requirement'

require 'beaker-puppet'

step 'Copy module to master VM' do
  metadata = JSON.parse(File.read('metadata.json'))
  module_tarball = "#{metadata['name']}-#{metadata['version']}.tar.gz"

  # We use this instead of `copy_module_to` from beaker-puppet as the
  # Rakefile contains logic for building a tarball that contains an
  # assembled JAR.
  master.do_scp_to("pkg/#{module_tarball}", "/tmp/#{module_tarball}", {})
  # The uninstall is here to ensure we load in a fresh copy of the
  # module if the tests are being re-run.
  on(master, puppet('module', 'uninstall', 'sharpie-facts_upload'),
    accept_all_exit_codes: true)
  on(master, puppet('module', 'install', "/tmp/#{module_tarball}"))

  # Trigger type generation and reload puppetserver to flush cached types.
  on(master, puppet('generate', 'types'))
  master_service = master[:type].start_with?('foss') ? 'puppetserver' : 'pe-puppetserver'
  bounce_service(master, master_service)
end

step 'Install module on master VM' do
  # The module currently supports Puppet Server 5.1 and 5.2 as native
  # support for uploading facts was added in 5.3.
  expect_service_mounted = if Gem::Requirement.new('< 5.3').satisfied_by?(Gem::Version.new(ENV['PUPPETSERVER_VERSION']))
                             true
                           else
                             false
                           end

  manifest = <<-EOM
if fact('pe_server_version') =~ String {
  # Pulls in other PE config bits that facts_upload::server hooks into.
  include puppet_enterprise::profile::certificate_authority
} else {
  service{'puppetserver': ensure => running}
}

class{'facts_upload::server': }
EOM
  apply_manifest_on(master, manifest)

  status_check = "curl -k https://127.0.0.1:8140/status/v1/services"
  response = JSON.parse(on(master, status_check).stdout.chomp)

  if expect_service_mounted
    refute_nil(response['facts-upload-service'])
  else
    assert_nil(response['facts-upload-service'])
  end
end


project.clj
(def puppetserver-version
  "Version of Puppet Server to develop and test against"
  (get (System/getenv) "PUPPETSERVER_VERSION" "5.1.6"))

(defproject sharpie/facts-upload "1.1.0"
  :description "Puppet Server endpoint for the facts upload command"
  :license {:name "Apache License 2.0"
            :url "http://www.apache.org/licenses/LICENSE-2.0.html"}

  :pedantic? :abort

  :min-lein-version "2.7.1"

  :plugins [[lein-parent "0.3.5"]]

  ; clj-parent version used by puppetserver 5.1.6 / PE 2017.3.10
  :parent-project {:coords [puppetlabs/clj-parent "1.7.1"]
                   :inherit [:managed-dependencies]}

  :source-paths ["src/clj"]
  :resource-paths ["src/ruby"]
  :test-paths ["test/integration"]

  :dependencies [[org.clojure/clojure]

                 [ring/ring-core]

                 [puppetlabs/comidi]
                 [puppetlabs/ring-middleware]

                 [puppetlabs/trapperkeeper]
                 [puppetlabs/trapperkeeper-status]
                 [puppetlabs/trapperkeeper-webserver-jetty9]

                 [puppetlabs/puppetserver ~puppetserver-version]]

  :profiles {:dev {:source-paths ["dev"]
                   :repl-options {:init-ns tk-devtools}
                   :resource-paths ["dev-resources"]
                   :dependencies [[org.clojure/tools.namespace]
                                  [org.clojure/tools.nrepl]

                                  [cheshire]
                                  [ring-mock]

                                  ;; Re-declare dependencies with "test"
                                  ;; classifiers to pull in additional testing
                                  ;; code, helper functions and libraries.
                                  [puppetlabs/trapperkeeper-webserver-jetty9 nil :classifier "test"]
                                  [puppetlabs/trapperkeeper nil :classifier "test" :scope "test"]
                                  [puppetlabs/kitchensink nil :classifier "test" :scope "test"]

                                  [puppetlabs/http-client]
                                  [me.raynes/fs]
                                  ;; Convert data between JRuby and Clojure objects.
                                  [zweikopf "1.0.2" :exclusions [org.jruby/jruby-complete]]

                                  [puppetlabs/puppetserver ~puppetserver-version :classifier "test" :scope "test"]]}

             :puppet-module {:jar-name "facts-upload.jar"}}

  :test-selectors {:integration :integration}

  :aliases {"tk" ["trampoline" "run"
                  "--bootstrap-config" "dev-resources/facts-upload/bootstrap.cfg"
                  "--config" "dev-resources/facts-upload/config.conf"]}

  :main puppetlabs.trapperkeeper.main)


dev/tk_devtools.clj
(ns tk-devtools
  (:require [clojure.repl :refer :all]
            [clojure.pprint :as pprint]
            [clojure.java.io :as io]
            [clojure.tools.namespace.repl :refer [set-refresh-dirs refresh]]
            [puppetlabs.trapperkeeper.app :as tka]
            [puppetlabs.trapperkeeper.bootstrap :as bootstrap]
            [puppetlabs.trapperkeeper.config :as config]
            [puppetlabs.trapperkeeper.core :as tk]))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;; Basic system life cycle

(def system nil)

;; Ensure only source files from the current project are refreshed. This
;; excludes the checkouts directory as puppetserver has tests containing
;; relative paths that don't resolve correctly.
(set-refresh-dirs "./src/clj" "./dev")

(def bootstrap-config
  (-> "facts-upload/bootstrap.cfg"
      io/resource
      .getPath))
(def app-config
  (-> "facts-upload/config.conf"
      io/resource
      .getPath))

(defn init []
  (alter-var-root #'system
                  (fn [_] (tk/build-app
                            (bootstrap/parse-bootstrap-config! bootstrap-config)
                            (config/load-config app-config))))
  (alter-var-root #'system tka/init)
  (tka/check-for-errors! system))

(defn start []
  (alter-var-root #'system
                  (fn [s] (if s (tka/start s))))
  (tka/check-for-errors! system))

(defn stop []
  (alter-var-root #'system
                  (fn [s] (when s (tka/stop s)))))

(defn go []
  (stop)
  (init)
  (start))

(defn reset []
  (stop)
  (refresh :after 'tk-devtools/go))

;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;;; Utilities for interacting with running system

(defn context
  "Get the current TK application context.  Accepts an optional array
  argument, which is treated as a sequence of keys to retrieve a nested
  subset of the map (a la `get-in`)."
  ([]
   (context []))
  ([keys]
   (get-in @(tka/app-context system) keys)))

(defn print-context
  "Pretty-print the current TK application context.  Accepts an optional
  array of keys (a la `get-in`) to print a nested subset of the context."
  ([]
   (print-context []))
  ([keys]
   (pprint/pprint (context keys))))


lib/puppet/face/facts/upload.rb
require 'puppet/face/facts'
require 'puppet/indirector/facts/rest'

unless  Puppet.features.facts_upload_builtin?
Puppet::Face.define(:facts, '0.0.1') do
  action(:upload) do
    summary "Upload local facts to the puppet master."
    description <<-'EOT'
      Reads facts from the local system using the `facter` terminus, then
      saves the returned facts using the rest terminus.
    EOT
    returns "Nothing."
    notes <<-'EOT'
      This action requires that the puppet master's `auth.conf` file
      allow save access to the `facts` REST terminus. Puppet agent does
      not use this facility, and it is turned off by default. See
      <https://puppet.com/docs/puppetserver/latest/config_file_auth.html>
      for more details.
    EOT
    examples <<-'EOT'
      Upload facts:

      $ puppet facts upload
    EOT

    when_invoked do |options|
      # Use `agent` sections  settings for certificates, Puppet Server URL,
      # etc. instead of `user` section settings.
      Puppet.settings.preferred_run_mode = :agent
      Puppet::Node::Facts.indirection.terminus_class = :facter

      facts = Puppet::Node::Facts.indirection.find(Puppet[:node_name_value])
      unless Puppet[:node_name_fact].empty?
        Puppet[:node_name_value] = facts.values[Puppet[:node_name_fact]]
        facts.name = Puppet[:node_name_value]
      end

      Puppet::Node::Facts.indirection.terminus_class = :rest
      server = Puppet::Node::Facts::Rest.server
      Puppet.notice "Uploading facts for '#{Puppet[:certname]}' to: '#{server}'"

      Puppet::Node::Facts.indirection.save(facts)

      nil
    end
  end
end
end


src/ruby/ruby/monkeypatches/facts_upload.rb
# Required to prevent the IndirectedRoutes class from being patched
# twice. Patching twice causes alias_method to set up a loop that
# leads to "stack level too deep" errors for non-Facts requests.
#
# A better way to do this would be to update the JRuby load path to
# include this directory and use a plain `require
# 'monkeypatches/facts_upload'` as require has logic to prevent a file
# from being evaluated twice.
if not defined?($_sharpie_facts_upload_patch)
  require 'puppet/indirector/facts/yaml'
  require 'puppet/network/http/api/indirected_routes'

  # NOTE: If JRuby 9k was in use, we could use Module#prepend from Ruby 2.x
  # which is a much nicer method of monkeypatching.

  class Puppet::Network::HTTP::API::IndirectedRoutes
    alias_method :_unpatched_plurality, :plurality

    def plurality(indirection)
      # This is required because at some point someone thought that making the
      # puppet HTTP handlers sensitive to inflection, but then centralizing all
      # the handling for that sensitivity into a static method was a good idea.
      #
      # I hope that someone catches a nasty case of the hiccoughs.
      return :singular if indirection == "facts"
      # Delegate to original method
      _unpatched_plurality(indirection)
    end
  end

  class Puppet::Node::Facts::Yaml
    def allow_remote_requests?
      true
    end
  end

  $_sharpie_facts_upload_patch = true
end


src/clj/puppetlabs/services/facts_upload/facts_upload_core.clj
(ns puppetlabs.services.facts-upload.facts-upload-core
  (:require
   [clojure.tools.logging :as log]
   [clojure.java.io :as io]
   [puppetlabs.trapperkeeper.services.status.status-core :as status-core]
   [puppetlabs.services.master.master-core :as master-core]
   [puppetlabs.services.jruby-pool-manager.jruby-core :as jruby-core]
   [puppetlabs.services.protocols.jruby-puppet :as jruby-puppet]
   [puppetlabs.services.request-handler.request-handler-core :as jruby-handler]
   [puppetlabs.puppetserver.jruby-request :as jruby-middleware]
   [puppetlabs.comidi :as comidi]
   [ring.util.response :as response]))

(def patchfile "ruby/monkeypatches/facts_upload.rb")

(def facts-upload-monkeypatch
  "Returns the path to a monkeypatch for re-enabling the facts terminus,
  if nothing terrible has happened to Java resource paths. Otherwise,
  returns nil which causes the wrap-with-monkeypatches handler to no-op."
  (when-let [monkeypatch (io/resource patchfile)]
    monkeypatch))


(defn create-status-callback
  "Creates a callback function for the trapperkeeper-status library.
  Used to report the presence of this Puppet Server extension."
  [context]
  (fn
    [level]
      (let [level>= (partial status-core/compare-levels >= level)]
        {:state :running
         :status {}})))

(defn wrap-with-monkeypatches
  "This middleware function ensures that the JRuby handling a request
  has been monkey-patched to re-enable handling of HTTP requests for
  the facts terminus."
  [handler]
  (if facts-upload-monkeypatch
    (fn [request]
      (when-let [jruby (:jruby-container request)]
        (.runScriptlet jruby
                       (io/input-stream facts-upload-monkeypatch)
                       (.getPath facts-upload-monkeypatch)))
      (handler request))
    (do
      (log/error (str "Unable to load facts_upload patch from: " patchfile))
      handler)))

(defn wrap-with-jruby-instance
  "A re-implementation of jruby-middleware/wrap-with-jruby-instance that
  exposes the scripting container in addition to the JRubyPuppet interface."
  [handler jruby-service]
  (let [jruby-pool (jruby-puppet/get-pool-context jruby-service)]
    (fn [request]
      (let [borrow-reason {:request (dissoc request :ssl-client-cert)}]
        (jruby-core/with-jruby-instance jruby-instance jruby-pool borrow-reason
          (-> request
              (assoc :jruby-instance (:jruby-puppet jruby-instance))
              (assoc :jruby-container (:scripting-container jruby-instance))
              handler))))))

(defn create-jruby-handler
  "Basically jruby-handler/build-request-handler with an injected monkeypatch."
  [jruby-service puppet-config code-id-fn]
  (let [config (jruby-handler/config->request-handler-settings puppet-config)]
    (-> (jruby-handler/jruby-request-handler config code-id-fn)
        wrap-with-monkeypatches ; <- This is the whole reason for most of the
                                ; code in this file...
        (wrap-with-jruby-instance jruby-service) ; <- This too.
        jruby-middleware/wrap-with-error-handling)))

(defn create-wrapped-jruby-handler
  "Builds a wrapped JRuby handler similar to that provided by
  the RequestHandlerService after being wrapped with authorization
  and other middleware from the MasterService.

  The reason we duplicate it here is so that we can add middleware
  that injects monkey-patches to re-activate the Ruby bits of
  the facts terminus."
  [puppet-config code-id-fn jruby-service auth-handler]
  (let [jruby-handler (create-jruby-handler jruby-service puppet-config code-id-fn)
        puppet-version (get-in puppet-config [:puppetserver :puppet-version])
        use-legacy-auth? (get-in puppet-config
                           [:jruby-puppet :use-legacy-auth-conf] false)]
    (master-core/get-wrapped-handler jruby-handler
                                     auth-handler
                                     puppet-version
                                     use-legacy-auth?)))

(defn create-request-routes
  "Builds a Comidi routing tree that responds to PUT requests for facts."
  [jruby-handler]
  (comidi/routes
    (comidi/context "/puppet/v3"
     (comidi/routes
      (comidi/PUT ["/facts/" [#".*" :rest]] request
                  (jruby-handler request))))))

(defn create-request-handler
  "Creates a Ring handler that responds to PUT requests for facts using
  a supplied JRuby handling function."
  [handler]
  (comidi/routes->handler
    (create-request-routes handler)))


src/clj/puppetlabs/services/facts_upload/facts_upload_service.clj
(ns puppetlabs.services.facts-upload.facts-upload-service
  (:require
   [clojure.edn :as edn]
   [clojure.string :as string]
   [clojure.tools.logging :as log]
   [puppetlabs.services.facts-upload.facts-upload-core :as core]
   [puppetlabs.trapperkeeper.core :refer [defservice]]
   [puppetlabs.trapperkeeper.services :as services]
   [puppetlabs.trapperkeeper.services.status.status-core :as status-core]))


(def version
  "The facts_upload module version, from project.clj"
  (status-core/get-artifact-version "sharpie" "facts-upload"))

(def puppetserver-version
  "Retrieve Puppet Server version in use.

  This function retrieves just the X and Y components of the version
  and casts them to a floating point number that can be compared."
  (try
    (as-> (status-core/get-artifact-version "puppetlabs" "puppetserver") x
          (string/split x #"\.")
          (take 2 x)
          (string/join "." x)
          (edn/read-string x))
    (catch Exception e
      (log/errorf "Couldn't determine Puppet Server version %s" e)
      nil)))

(defprotocol FactsUploadService)


(defn compatible-puppetserver-version?
  "Determine whether Puppet Server is compatible with facts_upload.

  This function returns a simple boolean value."
  []
  (if (nil? puppetserver-version)
    false
    (and (>= puppetserver-version 2.6)
         (< puppetserver-version 5.3))))

(defservice facts-upload-service
  FactsUploadService
  [[:StatusService register-status]
   [:WebserverService add-ring-handler]

   [:PuppetServerConfigService get-config]
   [:CaService get-auth-handler]
   [:VersionedCodeService current-code-id]
   ;; Mainly here to ensure we start after the MasterService
   [:MasterService]]
  (init [this context]
    ;; TODO: Only build the handler if we detect a compatible version of
    ;; Puppet Server.
    (if (compatible-puppetserver-version?)
      (do
        (log/info "Initializing FactsUpload service")
        (let [puppet-config (get-config)
              auth-handler (get-auth-handler)
              jruby-service (services/get-service this :JRubyPuppetService)
              jruby-handler (core/create-wrapped-jruby-handler puppet-config
                                                               current-code-id
                                                               jruby-service
                                                               auth-handler)
              facts-handler (core/create-request-handler jruby-handler)]

          (register-status "facts-upload-service" version 1
                           (core/create-status-callback context))

          (assoc context :request-handler facts-handler)))
      (do
        (log/warnf "The facts_upload plugin is not compatible with Puppet Server version %s and should be removed from this node. Skipping service initialization."
                    puppetserver-version)
        context)))


  (start [this context]
    ;; FIXME: Use the WebroutingService to lookup which path the MasterService
    ;; is mounted at.
    (when-let [handler (:request-handler context)]
      (log/info "Starting FactsUpload service")
      (add-ring-handler handler "/puppet/v3/facts"))

    context))
